src/Storages/MergeTree/MergeTreeDataPartWriterWide.cpp

```diff
@@ -172,43 +174,106 @@ void MergeTreeDataPartWriterWide::write(const Block & block, const IColumn::Perm
 
     auto granules_to_write = getGranulesToWrite(index_granularity, block.rows(), getCurrentMark(), rows_written_in_last_mark);
 
-    auto offset_columns = written_offset_columns ? *written_offset_columns : WrittenOffsetColumns{};
     Block primary_key_block;
     if (settings.rewrite_primary_key)
         primary_key_block = getBlockAndPermute(block, metadata_snapshot->getPrimaryKeyColumns(), permutation);
 
     Block skip_indexes_block = getBlockAndPermute(block, getSkipIndicesColumns(), permutation);
 
+    std::unique_ptr<ThreadPool> writing_thread_pool;
+    std::vector<WrittenOffsetColumns> offset_columns_per_column;
+    std::vector<bool> column_data_written(columns_list.size());
+    if (settings.max_threads_per_block != 1)
+    {
+        offset_columns_per_column.reserve(columns_list.size());
+        writing_thread_pool = std::make_unique<ThreadPool>(settings.max_threads_per_block);
+    }
+
+    offset_columns_per_column.emplace_back(written_offset_columns ? *written_offset_columns : WrittenOffsetColumns{});
+    if (writing_thread_pool)
+    {
+        auto it = columns_list.begin();
+        for (size_t i = 0; i + 1 < columns_list.size(); ++i, ++it)
+        {
+            const ColumnWithTypeAndName & column = block.getByName(it->name);
+            offset_columns_per_column.emplace_back(getOffsetColumnsForColumn(column.name, *column.type, offset_columns_per_column.back()));
+        }
+    }
+
     auto it = columns_list.begin();
     for (size_t i = 0; i < columns_list.size(); ++i, ++it)
     {
+        WrittenOffsetColumns & offset_columns = writing_thread_pool ? offset_columns_per_column[i] : offset_columns_per_column.back();
+        const ColumnWithTypeAndName & column = block.getByName(it->name);
+
+        prepareWriteColumn(column.name, *column.type, offset_columns);
+    }
+
+    it = columns_list.begin();
+    for (size_t i = 0; i < columns_list.size(); ++i, ++it)
```

Comment:
Parallelizing data write by columns might undermine the idea altogether in case of single fat column - parallelization would become completely pointless in such case.